File: LyllyPlayer/LyllyPlayer.Core/Services/PlayOrderService.cs
namespace LyllyPlayer.Services;

/// <summary>
/// Shuffle history tape + recently-played window (used for shuffle next/prev and candidate filtering).
/// </summary>
public sealed class PlayOrderService
{
    private readonly List<string> _shuffleTapeVideoIds = new();
    private readonly List<string> _recentlyPlayedVideoIds = new(5);
    private int _shuffleTapeCursor = -1;

    public IReadOnlyList<string> ShuffleTapeVideoIds => _shuffleTapeVideoIds;

    public int ShuffleTapeCursor => _shuffleTapeCursor;

    public int GetRecentlyPlayedWindowSize(int playlistEntryCount)
    {
        var n = Math.Max(0, playlistEntryCount);
        var scaled = (int)Math.Ceiling(Math.Sqrt(n));
        return Math.Clamp(scaled, 5, 40);
    }

    public int GetShuffleTapeMaxSize(int playlistEntryCount)
    {
        var n = Math.Max(0, playlistEntryCount);
        var scaled = (int)Math.Ceiling(n * 0.10);
        var max = Math.Clamp(scaled, 10, 200);
        return Math.Min(max, n == 0 ? max : n);
    }

    public void RecordNowPlayingForShuffleTape(string videoId, int playlistEntryCount)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(videoId))
                return;

            if (_shuffleTapeCursor >= 0 && _shuffleTapeCursor < _shuffleTapeVideoIds.Count)
            {
                if (string.Equals(_shuffleTapeVideoIds[_shuffleTapeCursor], videoId, StringComparison.OrdinalIgnoreCase))
                    return;

                if (_shuffleTapeCursor + 1 < _shuffleTapeVideoIds.Count &&
                    string.Equals(_shuffleTapeVideoIds[_shuffleTapeCursor + 1], videoId, StringComparison.OrdinalIgnoreCase))
                {
                    _shuffleTapeCursor++;
                    return;
                }

                if (_shuffleTapeCursor - 1 >= 0 &&
                    string.Equals(_shuffleTapeVideoIds[_shuffleTapeCursor - 1], videoId, StringComparison.OrdinalIgnoreCase))
                {
                    _shuffleTapeCursor--;
                    return;
                }

                if (_shuffleTapeCursor < _shuffleTapeVideoIds.Count - 1)
                    _shuffleTapeVideoIds.RemoveRange(_shuffleTapeCursor + 1, _shuffleTapeVideoIds.Count - (_shuffleTapeCursor + 1));
            }

            if (_shuffleTapeVideoIds.Count == 0 ||
                !string.Equals(_shuffleTapeVideoIds[^1], videoId, StringComparison.OrdinalIgnoreCase))
            {
                _shuffleTapeVideoIds.Add(videoId);
            }
            _shuffleTapeCursor = _shuffleTapeVideoIds.Count - 1;

            var max = GetShuffleTapeMaxSize(playlistEntryCount);
            while (_shuffleTapeVideoIds.Count > max)
            {
                _shuffleTapeVideoIds.RemoveAt(0);
                _shuffleTapeCursor--;
            }
            _shuffleTapeCursor = Math.Clamp(_shuffleTapeCursor, -1, _shuffleTapeVideoIds.Count - 1);
        }
        catch { /* ignore */ }
    }

    public void RecordRecentlyPlayedVideoId(string videoId, int playlistEntryCount)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(videoId))
                return;

            _recentlyPlayedVideoIds.Add(videoId);
            var max = GetRecentlyPlayedWindowSize(playlistEntryCount);
            while (_recentlyPlayedVideoIds.Count > max)
                _recentlyPlayedVideoIds.RemoveAt(0);
        }
        catch { /* ignore */ }
    }

    public bool RecentlyPlayedContains(string videoId) =>
        _recentlyPlayedVideoIds.Exists(x => string.Equals(x, videoId, StringComparison.OrdinalIgnoreCase));

    public void ClearRecentlyPlayed() => _recentlyPlayedVideoIds.Clear();

    public int RecentlyPlayedCount => _recentlyPlayedVideoIds.Count;

    public void ClearShuffleTape()
    {
        _shuffleTapeVideoIds.Clear();
        _shuffleTapeCursor = -1;
    }
}

File: android/app/src/main/java/com/lyllyplayer/app/playorder/PlayOrderService.kt
package com.lyllyplayer.app.playorder

import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Port of LyllyPlayer.Core PlayOrderService — shuffle tape + recently-played window.
 */
class PlayOrderService {
    private val shuffleTapeVideoIds = ArrayList<String>()
    private val recentlyPlayedVideoIds = ArrayList<String>(5)
    private var shuffleTapeCursor: Int = -1

    val shuffleTapeIds: List<String> get() = shuffleTapeVideoIds.toList()
    val tapeCursor: Int get() = shuffleTapeCursor
    val recentlyPlayedCount: Int get() = recentlyPlayedVideoIds.size

    fun getRecentlyPlayedWindowSize(playlistEntryCount: Int): Int {
        val n = max(0, playlistEntryCount)
        val scaled = ceil(sqrt(n.toDouble())).toInt()
        return scaled.coerceIn(5, 40)
    }

    fun getShuffleTapeMaxSize(playlistEntryCount: Int): Int {
        val n = max(0, playlistEntryCount)
        val scaled = ceil(n * 0.10).toInt()
        val maxSize = scaled.coerceIn(10, 200)
        return if (n == 0) maxSize else min(maxSize, n)
    }

    fun recordNowPlayingForShuffleTape(videoId: String, playlistEntryCount: Int) {
        if (videoId.isBlank()) return
        try {
            if (shuffleTapeCursor >= 0 && shuffleTapeCursor < shuffleTapeVideoIds.size) {
                if (shuffleTapeVideoIds[shuffleTapeCursor].equals(videoId, ignoreCase = true)) return

                if (shuffleTapeCursor + 1 < shuffleTapeVideoIds.size &&
                    shuffleTapeVideoIds[shuffleTapeCursor + 1].equals(videoId, ignoreCase = true)
                ) {
                    shuffleTapeCursor++
                    return
                }

                if (shuffleTapeCursor - 1 >= 0 &&
                    shuffleTapeVideoIds[shuffleTapeCursor - 1].equals(videoId, ignoreCase = true)
                ) {
                    shuffleTapeCursor--
                    return
                }

                if (shuffleTapeCursor < shuffleTapeVideoIds.size - 1) {
                    shuffleTapeVideoIds.subList(shuffleTapeCursor + 1, shuffleTapeVideoIds.size).clear()
                }
            }

            if (shuffleTapeVideoIds.isEmpty() ||
                !shuffleTapeVideoIds.last().equals(videoId, ignoreCase = true)
            ) {
                shuffleTapeVideoIds.add(videoId)
            }
            shuffleTapeCursor = shuffleTapeVideoIds.lastIndex

            val maxSize = getShuffleTapeMaxSize(playlistEntryCount)
            while (shuffleTapeVideoIds.size > maxSize) {
                shuffleTapeVideoIds.removeAt(0)
                shuffleTapeCursor--
            }
            shuffleTapeCursor = shuffleTapeCursor.coerceIn(-1, shuffleTapeVideoIds.lastIndex)
        } catch (_: Exception) {
            // ignore
        }
    }

    fun recordRecentlyPlayedVideoId(videoId: String, playlistEntryCount: Int) {
        if (videoId.isBlank()) return
        try {
            recentlyPlayedVideoIds.add(videoId)
            val maxSize = getRecentlyPlayedWindowSize(playlistEntryCount)
            while (recentlyPlayedVideoIds.size > maxSize) {
                recentlyPlayedVideoIds.removeAt(0)
            }
        } catch (_: Exception) {
            // ignore
        }
    }

    fun recentlyPlayedContains(videoId: String): Boolean =
        recentlyPlayedVideoIds.any { it.equals(videoId, ignoreCase = true) }

    fun clearRecentlyPlayed() = recentlyPlayedVideoIds.clear()

    fun clearShuffleTape() {
        shuffleTapeVideoIds.clear()
        shuffleTapeCursor = -1
    }
}
